### PR TITLE
Fix SortPreservingRepartition with no existing ordering.

### PR DESCRIPTION
**Pull request by mustafasrepo**
## Which issue does this PR close?

Closes [#7794](https://github.com/apache/arrow-datafusion/issues/7794).

## What changes are included in this PR?

In the issue, `EnforceSorting` moves up the sort that satisfies the window. In this case, operators that are below the `SortExec` do not have ordering at their inputs. However, some of them (`SortPreservingRepartitionExec`) expects its input to be ordered. While removing `SortExec` down the plan, we also replace operators with their corresponding variants that do not preserve order. With this change I believe `SortPreservingRepartitionExec` will be replaced with `RepartitionExec` in the issue.

## Are these changes tested?
Yes, a new test is added to replicate problem in the issue.

### CODE REVIEW

File: datafusion/core/src/physical_optimizer/enforce_sorting.rs [hunk 1]
```diff
@@ -1635,14 +1645,16 @@ mod tests {
         // During the removal of `SortExec`s, it should be able to remove the
         // corresponding SortExecs together. Also, the inputs of these `SortExec`s
         // are not necessarily the same to be able to remove them.
-        let expected_input = ["BoundedWindowAggExec: wdw=[count: Ok(Field { name: \"count\", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), frame: WindowFrame { units: Range, start_bound: Preceding(NULL), end_bound: CurrentRow }], mode=[Sorted]",
+        let expected_input = [
```
Comment (alamb) [inline datafusion/core/src/physical_optimizer/enforce_sorting.rs:1648]: The diff made it hard to see -- but I am pretty sure this test did not change, just the formatting did. Is that correct?

File: datafusion/core/src/physical_optimizer/enforce_sorting.rs [hunk 2]
```diff
@@ -2234,4 +2246,36 @@ mod tests {
...
+
+        let expected_input = [
+            "BoundedWindowAggExec: wdw=[count: Ok(Field { name: \"count\", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), frame: WindowFrame { units: Range, start_bound: Preceding(NULL), end_bound: CurrentRow }], mode=[Sorted]",
+            "  SortPreservingMergeExec: [a@0 ASC,b@1 ASC]",
+            "    SortPreservingRepartitionExec: partitioning=RoundRobinBatch(10), input_partitions=10",
+            "      RepartitionExec: partitioning=RoundRobinBatch(10), input_partitions=1",
+            "        SortExec: expr=[a@0 ASC,b@1 ASC]",
+            "          CsvExec: file_groups={1 group: [[x]]}, projection=[a, b, c, d, e], has_header=false",
+        ];
+        let expected_optimized = [
```
Comment (alamb) [inline datafusion/core/src/physical_optimizer/enforce_sorting.rs:2270]: Without the code in this PR, I verified this test failed. The plan showed the same pattern as described in https://github.com/apache/arrow-datafusion/issues/7794

```
    "BoundedWindowAggExec: wdw=[count: Ok(Field { name: \"count\", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), frame: WindowFrame { units: Range, start_bound: Preceding(NULL), end_bound: CurrentRow }], mode=[Sorted]",
    "  SortPreservingMergeExec: []",
    "    SortExec: expr=[a@0 ASC,b@1 ASC]",
    "      SortPreservingRepartitionExec: partitioning=RoundRobinBatch(10), input_partitions=10",
    "        RepartitionExec: partitioning=RoundRobinBatch(10), input_partitions=1",
    "          CsvExec: file_groups={1 group: [[x]]}, projection=[a, b, c, d, e], has_header=false",
```

**Review by alamb — APPROVED** [review]
Thank you very much @mustafasrepo  -- this would have taken me quite a while to figure out.

I reviewed this PR carefully and verified it solved my problem downstream.

**Comment by alamb** [comment]
I don't see any reason to delay merging this fix. Thanks again @mustafasrepo

File: datafusion/core/src/physical_optimizer/enforce_sorting.rs (hunk 1 above)
Comment (mustafasrepo) [reply datafusion/core/src/physical_optimizer/enforce_sorting.rs:1648]: Yes, just inserted new line

File: datafusion/core/src/physical_optimizer/enforce_sorting.rs [hunk 3]
```diff
@@ -1635,14 +1645,16 @@ mod tests {
...
+        let expected_input = [
+            "BoundedWindowAggExec: wdw=[count: Ok(Field { name: \"count\", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), frame: WindowFrame { units: Range, start_bound: Preceding(NULL), end_bound: CurrentRow }], mode=[Sorted]",
             "  SortPreservingMergeExec: [nullable_col@0 DESC NULLS LAST]",
             "    UnionExec",
             "      SortExec: expr=[nullable_col@0 DESC NULLS LAST]",
             "        ParquetExec: file_groups={1 group: [[x]]}, projection=[nullable_col, non_nullable_col], output_ordering=[nullable_col@0 ASC, non_nullable_col@1 ASC]",
             "      SortExec: expr=[nullable_col@0 DESC NULLS LAST]",
             "        ParquetExec: file_groups={1 group: [[x]]}, projection=[nullable_col, non_nullable_col], output_ordering=[nullable_col@0 ASC]"];
-        let expected_optimized = ["WindowAggExec: wdw=[count: Ok(Field { name: \"count\", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} }), frame: WindowFrame { units: Range, start_bound: CurrentRow, end_bound: Following(NULL) }]",
+        let expected_optimized = [
```
Comment (mustafasrepo) [inline datafusion/core/src/physical_optimizer/enforce_sorting.rs:1656]: similar to above case

**Pull request merged**